### PR TITLE
infra: update diff report file download to use GitHub token and curl

### PR DESCRIPTION
**Pull request by nrmancuso**
Noticed at https://github.com/checkstyle/checkstyle/pull/12979#issuecomment-1509958270 and elsewhere, it looks like we are failing to download files for report generation. Example from logs:
```
...
2023-04-15T20:33:26.0038234Z   PATCH_CONFIG_LINK: 
2023-04-15T20:33:26.0038722Z   LINK_FROM_PR: https://gist.githubusercontent.com/0xbakry/7853843aa3d69aee301d7a8d62d08ebb/raw/65a7d6bee0ad8ad789c17f778ce7d8b0056ae28b/my_projects
2023-04-15T20:33:26.0039194Z ##[endgroup]
2023-04-15T20:33:26.1529944Z ##[error]Process completed with exit code 8.
...

```

exit code 8 from wget typically denotes a error response from the server, read more at https://stackoverflow.com/questions/65677320/what-does-server-issued-an-error-responseexit-status-8-mean-in-wget#:~:text=The%20man%20page%20of%20wget,types%20of%20errors%20are%20encountered..

We really need to begin to migrate complex steps for diff report generation to a script so that we can take advantage of our regular shellcheck exeuction (instead of built in actionlint one) and clean up the yaml. Issue is opened at https://github.com/checkstyle/checkstyle/issues/12991

-----------------------------------
Tested locally:
```
➜  checkstyle git:(fix-diff-report-action) export DEFAULT_PROJECTS_LINK=https://raw.githubusercontent.com/checkstyle/contribution/master/checkstyle-tester/projects-to-test-on-for-github-action.properties

➜  checkstyle git:(fix-diff-report-action) export DIFF_CONFIG_LINK=https://gist.githubusercontent.com/0xbakry/a1fcb13eef42f29de8ff500892f34166/raw/1b0af8d2da606b273e8ef9aaed6278020f876202/config_check.xml

➜  checkstyle git:(fix-diff-report-action) export LINK_FROM_PR=https://gist.githubusercontent.com/0xbakry/7853843aa3d69aee301d7a8d62d08ebb/raw/65a7d6bee0ad8ad789c17f778ce7d8b0056ae28b/my_projects

➜  checkstyle git:(fix-diff-report-action) ./.ci/diff-report.sh download-files
Downloading files...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   983  100   983    0     0    496      0  0:00:01  0:00:01 --:--:--   496
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1278  100  1278    0     0    955      0  0:00:01  0:00:01 --:--:--   955

➜  checkstyle git:(fix-diff-report-action) ✗ ls
appveyor.yml            config           LICENSE           project.properties  SECURITY.md
azure-pipelines.yml     diff_config.xml  LICENSE.apache20  README.md           src
cdg-pitest-licence.txt  gen              pom.xml           RIGHTS.antlr        target

➜  checkstyle git:(fix-diff-report-action) ✗ cat project.properties 
# List of GIT repositories to clone / pull for checking with Checkstyle
# File format: REPO_NAME|[local|git|hg]|URL|[COMMIT_ID]|[EXCLUDE FOLDERS]
# Please note that bash comments works in this file

guava|git|https://github.com/google/guava|v28.2||

infinispan|git|https://github.com/infinispan/infinispan|7.2.5.Final||
protonpack|git|https://github.com/poetix/protonpack|protonpack-1.7||
jOOL|git|https://github.com/jOOQ/jOOL|version-0.9.7||
lucene-solr|git|https://github.com/apache/lucene-solr|releases/lucene-solr/5.4.0||

tapestry-5|git|https://github.com/apache/tapestry-5|5.4.0|**/archetype-resources/**/*|
...
```

Note that we have set this report to be generated with a custom projects file, this is why i `cat`'d it to show that script works.

### CODE REVIEW

**Comment by nrmancuso** [comment]
We can update branch and try to regenerate report at https://github.com/checkstyle/checkstyle/pull/12979 after we merge this to test.

File: .ci/diff-report.sh
Comment (romani) [inline .ci/diff-report.sh:39]: You need to pass it from GitHub action config

https://github.com/checkstyle/checkstyle/blob/82cdd1ed0dac91e7cfa749fca161df69d51949e2/.github/workflows/release-new-milestone-and-issues-in-other-repos.yml#L51

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: .ci/diff-report.sh
Comment (nrmancuso) [reply .ci/diff-report.sh:39]: Followed format from elsewhere in this file with uppercase GITHUB_TOKEN

**Review by romani — APPROVED** [review]

**Comment by romani** [comment]
Let's merge and test it sooner

**Comment by nrmancuso** [comment]
>
Looks like we need action to allow auto merge if CI pass :). To not watch PR state.


Like if a maintainer comments "ok to merge" we merge after CI is in green state?

**Comment by romani** [comment]
`Github, merge if CI has 109 passes`

**Comment by 0xbakry** [comment]
@nrmancuso Please can you explain this to me, is something like `LINK_FROM_PR` [diff-report.yml](https://github.com/checkstyle/checkstyle/pull/12990/files#diff-215466a0cdca12d3b615e7955c418897165026b9d8ad1ee047235db8441b02b8) global? Since you use it in [diff-report.sh](https://github.com/checkstyle/checkstyle/pull/12990/files#diff-321a8a97fe5fa4d5093e76c46a6217ac297e238203b485f6006f7f69f203025c) and I'm little confused how? 
I'm trying to solve https://github.com/checkstyle/checkstyle/issues/12991 and just wonder if all variables are seen in both files.

**Comment by nrmancuso** [comment]
>I'm trying to solve https://github.com/checkstyle/checkstyle/issues/12991 and just wonder if all variables are seen in both files.

I just removed approved label, we need to do some more planning before working on this. The script you are asking about is ran from the yml file where the `LINK_FROM_PR` variable is defined.

**Comment by 0xbakry** [comment]
My bad, I missed it was defined as `environment variable`, thanks.
Do you suggest me if I can work on it now?

**Comment by nrmancuso** [comment]
@0xbakry I had planned on taking this one, it might be easier for a maintainer to complete this issue, I have already completed some of the work on this in a personal repo. Let's focus on getting https://github.com/checkstyle/checkstyle/pull/12979 ready for review.

**Pull request merged**
